### PR TITLE
Make CiliumNetworkPolicy CR creation be deployed or not with a flag

### PR DESCRIPTION
**Pull request by whites11**
How it was before, it you install external-dns app into a cluster with cilium before cilium is installed, you end up NOT having the CiliumNetworkPolicy created and the app being non functioning.

With this PR, we make control if the CR should be created or not using a flag in the values (default false).
If the flag is true and external-dns gets installed before cilium, installation will fail and retried.

WARNING: this is a breaking change

### CODE REVIEW

**Comment by Gacko** [comment]
Gonna review (and align to other manifests where required) today.

File: helm/external-dns-app/templates/ciliumnetworkpolicy.yaml
Comment (mcharriere) [inline helm/external-dns-app/templates/ciliumnetworkpolicy.yaml:1]: Is there any reason to move this resource outside the `crd` directory? as far as I know, there is no change in behavior. 
If that's the case, I'd suggest to move it back to the crd jobs folder to maintain them separated (eventually we'd like to move the crd job to an chart dependency).

File: helm/external-dns-app/templates/crd/np.yaml
Comment (mcharriere) [inline helm/external-dns-app/templates/crd/np.yaml:28]: this else was here to install one policy or the other one. I don't think makes sense to keep them both, as they are kind of doing the same thing.

**Review by mcharriere — CHANGES_REQUESTED** [review]
> WARNING: this is a breaking change

could you explain to what extend?

**Comment by whites11** [comment]
> > WARNING: this is a breaking change
> 
> could you explain to what extend?

before my change, the `cilium network policy` was created on a best effort/random fashion.
There were basically 2 possibilities:

- external-dns app installed before cilium app -> no cilium network policy
- external-dns app installed after cilium app -> cilium network policy created

with this change, the cilium network policy creation is not attempted at all, unless the `ciliumNetworkPolicy.enabled` flag gets set to true. This means users of this app have to flip that flag to get something close to the old behaviour.

does it make more sense?

**Comment by mcharriere** [comment]
> with this change, the cilium network policy creation is not attempted at all, unless the ciliumNetworkPolicy.enabled flag gets set to true. This means users of this app have to flip that flag to get something close to the old behaviour.

> does it make more sense?

yes it does, thank you. I was asking because we are preparing a v3 for this app with breaking changes. I don't think it makes sense to hold this change until v3, announcing it well should be enough.

File: helm/external-dns-app/templates/ciliumnetworkpolicy.yaml
Comment (whites11) [reply helm/external-dns-app/templates/ciliumnetworkpolicy.yaml]: good point, done

File: helm/external-dns-app/templates/crd/np.yaml
Comment (whites11) [reply helm/external-dns-app/templates/crd/np.yaml:28]: it still makes sense IMHO.
It is true that both policies have the same behaviour/meaning, but the `cilium` one is being added only because there is an upstream bug that prevents our use case from being addressable using k8s network policies only.
We would have preferred to avoid this vendor lock-in but we just can't.

An alternative solution would be to get rid of the cilium network policy and have an allow-all egress policy for the crd job. since this is a temporary job, the tradeoff in terms of security should not be too much.

WDYT? would that be better?

Comment (mcharriere) [reply helm/external-dns-app/templates/crd/np.yaml:28]: > upstream bug that prevents our use case from being addressable using k8s network policies only

ah I see, I was not aware of that. I'm fine with keeping both until we resolve the vendor lock-in in the future

**Review by mcharriere — APPROVED** [review]
looks good to me!

**Comment by whites11** [comment]
@Gacko you happy or want me to wait before merging?

**Comment by Gacko** [comment]
Yeah, go one. Since we are aligning this anyway, I don't think it makes sense to put more work into this as long as it works.

**Pull request merged**
